src/hdmf/container.py

```diff
@@ -830,3 +833,218 @@ def __build_conf_methods(cls, conf_dict, conf_index, multi):
+            return self.data[idx][col]
+        elif isinstance(args, str):
+            col = self.__col_index.get(args)
+            if col is None:
+                raise KeyError(args)
+            return [row[col] for row in self.data]
+        else:
+            return self.data[idx]
+
+    def to_dataframe(self):
+        '''Produce a pandas DataFrame containing this table's data.
+        '''
+
+        data = {colname: self[colname] for ii, colname in enumerate(self.columns)}
+        return pd.DataFrame(data)
+
+    @classmethod
+    @docval(
+        {'name': 'df', 'type': pd.DataFrame, 'doc': 'input data'},
+        {'name': 'name', 'type': str, 'doc': 'the name of this container', 'default': None},
+        {
+            'name': 'extra_ok',
+            'type': bool,
+            'doc': 'accept (and ignore) unexpected columns on the input dataframe',
+            'default': False
+        },
+    )
+    def from_dataframe(cls, **kwargs):
+        '''Construct an instance of Table (or a subclass) from a pandas DataFrame. The columns of the dataframe
+        should match the columns defined on the Table subclass.
+        '''
+
+        df, name, extra_ok = getargs('df', 'name', 'extra_ok', kwargs)
+
+        cls_cols = list([col['name'] for col in getattr(cls, '__columns__')])
+        df_cols = list(df.columns)
+
+        missing_columns = set(cls_cols) - set(df_cols)
+        extra_columns = set(df_cols) - set(cls_cols)
+
+        if extra_columns:
+            raise ValueError(
+                'unrecognized column(s) {} for table class {} (columns {})'.format(
+                    extra_columns, cls.__name__, cls_cols
+                )
+            )
+
+        use_index = False
+        if len(missing_columns) == 1 and list(missing_columns)[0] == df.index.name:
+            use_index = True
+
+        elif missing_columns:
+            raise ValueError(
+                'missing column(s) {} for table class {} (columns {}, provided {})'.format(
+                    missing_columns, cls.__name__, cls_cols, df_cols
+                )
+            )
+
+        data = []
+        for index, row in df.iterrows():
+            if use_index:
+                data.append([
+                    row[colname] if colname != df.index.name else index
+                    for colname in cls_cols
+                ])
+            else:
+                data.append([row[colname] for colname in cls_cols])
+
+        if name is None:
+            return cls(data=data)
+        return cls(name=name, data=data)
+
+
+class TableRegion(DataRegion):
+    '''
+    A class for representing regions i.e. slices or indices into an Table
+    '''
+
+    @docval({'name': 'name', 'type': str, 'doc': 'the name of this container'},
+            {'name': 'table', 'type': Table, 'doc': 'the Table this region applies to'},
+            {'name': 'region', 'type': (slice, list, tuple, RegionReference), 'doc': 'the indices of the table'})
```

Comment:
Why do we need to HDF5 region references here? The reason I am asking it that this seems to make the class depended on the I/O backend.

Reply:
I think the idea was so you could create a `TableRegion` from an HDF5 region reference. This was pulled over from pynwb and was written years ago. I'm not sure we need it. The `get_region_slicer` function creates an import problem, too. So I'd be happy to axe the whole thing :)

Reply:
It looks like the class is not being used right now. As such, I think it is reasonable to remove it for now until we get an actual need.